src/controllers/engine/controllerengine.cpp

```diff
@@ -223,6 +229,51 @@ void ControllerEngine::uninitializeScriptEngine() {
     }
 }
 
+void ControllerEngine::loadModule(QFileInfo moduleFileInfo) {
+    m_moduleFileInfo = moduleFileInfo;
+
+    QJSValue mod = m_pScriptEngine->importModule(moduleFileInfo.absoluteFilePath());
+    if (mod.isError()) {
+        showScriptExceptionDialog(mod);
+        return;
+    }
+
+    connect(&m_scriptWatcher,
+            &QFileSystemWatcher::fileChanged,
+            this,
+            &ControllerEngine::scriptHasChanged);
+    m_scriptWatcher.addPath(moduleFileInfo.absoluteFilePath());
+
+    QJSValue initFunction = mod.property("init");
+    executeFunction(initFunction, QJSValueList{});
+
+    QJSValue handleInputFunction = mod.property("handleInput");
+    if (handleInputFunction.isCallable()) {
+        m_handleInputFunction = handleInputFunction;
+    } else {
+        scriptErrorDialog(
+                "Controller JavaScript module exports no handleInput function.",
+                QStringLiteral("handleInput"),
+                true);
+    }
+
+    QJSValue shutdownFunction = mod.property("shutdown");
+    if (shutdownFunction.isCallable()) {
+        m_shutdownFunction = shutdownFunction;
+    } else {
+        qDebug() << "Module exports no shutdown function.";
+    }
+}
+
+void ControllerEngine::handleInput(QByteArray data, mixxx::Duration timestamp) {
+    if (m_handleInputFunction.isCallable()) {
+        QJSValueList args;
+        args << byteArrayToScriptValue(data);
+        args << timestamp.toDoubleNanos();
```

Comment:
@daschuer, @Pegasus-RPG: What is an appropriate level of precision for this? I used the most precise that the Duration class offers but I don't know the timestamps are accurate to that level.

Reply:
ping @daschuer

Reply:
In case of PortMidi it has a ms resolution.

Reply:
Okay, I changed it to millisecond resolution in 359654e.